### ISSUE TITLE
[Proposal] check variable usage before checking sort order

### ISSUE THREAD

**Issue by coolaj86** [dotenv-linter/dotenv-linter#360]
In cases where a variable is used, the linter should not complain if it is not listed alphabetically.

This is correct:

```bash
FOO=something
BAR="$FOO or other"
```

This is NOT correct:

```bash
BAR="$FOO or other"
FOO=something
```

Workaround:

- Add a comment

```bash
FOO=something

# BAR uses FOO
BAR="$FOO or other"
```

**Comment by DDtKey** [comment]
This really sounds useful, as reordering of references can lead to incorrect values 🤔 



At the moment, the user can independently add a separator for highlighting the sorting groups:
```sh
FOO=something

BAR="$FOO or other"
```
Or use the [control comment to skip a check](https://dotenv-linter.github.io/#/usage/check?id=skip-checks):
```sh
# dotenv-linter:off UnorderedKey (You can disable a check for only some lines)
FOO=something
BAR="$FOO or other"
# dotenv-linter:on UnorderedKey (And enable it again)
```

**Comment by mgrachev** [comment]
I like this proposal 👍

@coolaj86 Which kinds of variable references should `UnorderedKeyChecker` support?

**Comment by Spoonbender** [comment]
I imagine that theimplementation would rely on something like a directed graph, which can also be used to detect cyclic dependencies

**Comment by WLBF** [comment]
Does dotenv have formal standard for variable reference?

**Comment by mdaj06** [comment]
@mgrachev can i take a look at this?

**Comment by mgrachev** [comment]
> Does dotenv have formal standard for variable reference?

I don't think there is a standard for this, but I know of some dotenv libraries that support variable substitution:
* https://github.com/dotenv-rs/dotenv#variable-substitution
* https://github.com/bkeepers/dotenv#variable-substitution
* https://github.com/vlucas/phpdotenv#nesting-variables
* https://github.com/stackbuilders/dotenv-hs#variable-substitution
* https://github.com/theskumar/python-dotenv#getting-started

> @mgrachev can i take a look at this?

Sure 🙂

**Comment by mgrachev** [comment]
@mdaj06 Do you have any progress on this?

**Comment by mgrachev** [comment]
We decomposed this issue for 3 parts:
* Add method to get substitution key to LineEntry (https://github.com/dotenv-linter/dotenv-linter/issues/379)
* Allow unordered keys for substituion variables (https://github.com/dotenv-linter/dotenv-linter/issues/382)
* Don't fix key order for substitution variables (https://github.com/dotenv-linter/dotenv-linter/issues/383)

Thanks to @DDtKey for that 👍

I'll close this issue.
